### ISSUE TITLE
[Feature Request]: Custom-Built Exponent Client

### ISSUE THREAD

**Issue by ericnograles** [expo/expo#86]
Hi guys,

First off, love what you've done.  Expo is fantastic and will tremendously help get React Native in the hands of folks who otherwise would not have access to it.  I'd be remiss to not also mention its kickass developer experience.  So, high fives all around!

I noticed your tooling for the Expo Client is _very_ close to how AppGyver implements their [Scanner](https://play.google.com/store/apps/details?id=com.appgyver.freshandroid&hl=en).  Same kind of concept -- the client already has the native entitlements the SDK expects, local dev environment has a utility to expose a port and QR code, the client QR code scan brings JS in, the app runs, and choirs of angels sing.

I saw that if you `detach` your native projects, it's basically back to "vanilla" React Native dev...which isn't terrible, but the Expo DX is so good that it will definitely feel like a downgrade.

I'm not sure if you guys have considered this, but AppGyver gave a "semi-escape hatch" for their Scanner.  Basically, you could build build custom versions of the scanner (in this case, the Expo Client) with whatever custom native libraries they have linked.  This would allow someone the latitude of "ejecting" and linking non-SDK libraries, but still have the DX of the AG Scanner.  

Now, if I recall, it was definitely a little involved, as you basically had to create a distribution profile for iOS and keystore for Android, upload it to AG's infrastructure, and then have it build your Custom Scanner.  I feel like you could even automate that process, since I see the `exp` cmd line utility creates that stuff for me.

Anyway, TL;DR: A custom-built Exponent Client would be an amazing future feature for Expo, as there's definitely a lot of use cases that would require ejecting out of the "vanilla SDK".

Thanks again for your work, and I'm looking forward to using (and hopefully contributing back to) this ecosystem!

Eric

**Comment by sonaye** [comment]
As far as I know, when ejecting from Expo, you still have access to all the goodies offered by Expo SDK, you only lose the build-related features, without ejecting publishing directly to the store is automated. So, not back to Vanilla RN for sure. I think this is ought to be clarified with the upcoming release this week in the docs section.

**Comment by jesseruder** [comment]
Correct @sonaye when you eject to ExpoKit, you keep almost all of the features offered by the Expo SDK. There are a few small things that are different right now - for example push notification don't work out of the box on Android.

@ericnograles does that make sense? Which part of the docs are confusing?

**Comment by ericnograles** [comment]
@jesseruder @sonaye totally makes sense, thanks.  And that's really what I was referring to was the awesomeness that is the Build Tooling.  To clarify, upon ejecting, you basically just have to manage it like a normal React Native app (Xcode/Android Studio) and can no longer do Expo.io Client magic, right?

**Comment by jesseruder** [comment]
Yes you need to use Xcode/Android Studio after detaching. Unfortunately not much we can do there, just a technical limitation.

**Comment by dikaiosune** [comment]
Hi! We're tracking feature requests on Canny now, and I've [created one there](https://expo.canny.io/feature-requests/p/build-a-custom-client-app-with-custom-native-modules) for this idea. Thanks!
